### PR TITLE
add expire_at method

### PR DESCRIPTION
**Pull request by cashaev**
Adding support for EXPIREAT key timestamp method in DefaultClient and ShardedClient.

fixes: #360

### CODE REVIEW

**Comment by WisdomPill** [comment]
Looks okay, 
I think that this method and some others need to give back the result of the operation (a boolean).

It is my first PR review, so I ask @jdufresne to confirm if everything looks okay.

**Comment by cashaev** [comment]
> Looks okay,
> I think that this method and some others need to give back the result of the operation (a boolean).
> 
> It is my first PR review, so I ask @jdufresne to confirm if everything looks okay.

Maybe so. But adding the result should definitely be consistent across all the methods, and possibly done in a separate PR?
Otherwise, this PR's scope will expand too much.
@WisdomPill Thank you very much for a prompt reply!

**Review by WisdomPill — COMMENTED** [review]
Thanks for solving #471 as well

**Review by WisdomPill — APPROVED** [review]
Looks good to me

**Comment by WisdomPill** [comment]
Hello @cashaev can you rebase on master so that we can merge this?

Thanks

**Comment by cashaev** [comment]
Any action on this PR? I need this functionality again now )))

**Comment by cashaev** [comment]
> Hello @cashaev can you rebase on master so that we can merge this?
> 
> Thanks

let me try

**Comment by cashaev** [comment]
> Hello @cashaev can you rebase on master so that we can merge this?
> 
> Thanks

Sorry, not sure if I can do it correctly. Any assistance? Do I need to reopen it from my master branch?

**Comment by jdufresne** [comment]
Please rebase. This should not contain changes to `.github/workflows/`, but they current show up in the diff.

**Comment by WisdomPill** [comment]
@cashaev let me explain the steps to rebase on `django-redis/master`.

- locally add another remote pointing to `jazzband/django-redis` with the following command: `git remote add upstream git@github.com:jazzband/django-redis.git`
- fetch from remote master `git fetch upstream master`
- make sure to be on your master with the following command: `git checkout master`. Or just check with `git status`
- rebase your master (`origin/master`) on jazzband's master with the following command: `git rebase upstream/master`
- rebase your feature branch on your own master with the following two commands `git checkout expireat` and `git rebase origin/master`
- if there will be a need to merge and stuff you will get prompts, after that you can push with `git push --force origin expireat`, force is needed because you are changing the history of you branch.

Theoretically you could skip rebasing your master on upstream/master, but it is just good practice to keep everything in sync.

Hope this helps, let me know if you need more assistance

**Review by terencehonles — APPROVED** [review]
LGTM, but I'm not sure if we want to be implementing _every_ redis command in this class. @WisdomPill thoughts? You want to squash this if you're good with it so we can include it in the next release?

**Comment by WisdomPill** [comment]
I think it is just a utility method recycling `.expire` method so it is okay to add it, I am worried that many commands are implemented only in `DefaultClient` so I think other clients could use some enhancements to be able to do the same things as `DefaultClient`.

File: tests/test_backend.py [hunk 1]
```diff
@@ -661,6 +661,31 @@ def test_expire(self):
...
+        self.assertAlmostEqual(ttl, timedelta(hours=1).total_seconds(), delta=1)
+
+        # Test settings expiration time 1 hour ahead by Unix timestamp.
+        self.cache.set("foo", "bar", timeout=None)
+        expiration_time = datetime.datetime.now() + timedelta(hours=2)
+        self.cache.expire_at("foo", int(expiration_time.timestamp()))
+        ttl = self.cache.ttl("foo")
+        self.assertAlmostEqual(ttl, timedelta(hours=1).total_seconds() * 2, delta=1)
+
+        """
```
Comment (WisdomPill) [inline tests/test_backend.py:680]: Can this be a multiline comment?

Comment (WisdomPill) [inline tests/test_backend.py:680]:
```suggestion
        # Test settings expiration time 1 hour in past, which
        # effectively deletes the key.
        
```

File: django_redis/client/default.py [hunk 2]
```diff
@@ -289,6 +290,19 @@ def expire(
         if client.exists(key):
             client.expire(key, timeout)
 
+    def expire_at(self, key, when: Union[datetime, int], version=None, client=None):
```
Comment (WisdomPill) [inline django_redis/client/default.py:293]:
```suggestion
    def expire_at(self, key: Any, when: Union[datetime, int], version: Optional[int]=None, client: Optional[Redis]=None):
```

**Review by WisdomPill — CHANGES_REQUESTED** [review]

**Comment by terencehonles** [comment]
@WisdomPill I just applied your suggestions myself. If you're good with it as is, I'll let you handle squashing this.

**Review by WisdomPill — APPROVED** [review]

**Comment by WisdomPill** [comment]
Thanks @terencehonles!

**Comment by WisdomPill** [comment]
wait a mement

**Review by WisdomPill — CHANGES_REQUESTED** [review]
`when` can be a `datetime` but using a date time will get an exception, `datetime` needs to be converted to unix timestamp

**Comment by terencehonles** [comment]
> `when` can be a `datetime` but using a date time will get an exception, `datetime` needs to be converted to unix timestamp

Where? The tests cover sending both a `datetime` and an `int`, so that's covered there. The underlying client (if `redis-py`) handles converting to an `int` https://github.com/andymccurdy/redis-py/blob/1a41cfd95a53a95b078084d8627be6b6fba3bb71/redis/client.py#L1659-L1660

**Comment by WisdomPill** [comment]
Okay, sorry did not notice that redis py supports it, my bad

**Pull request merged**